### PR TITLE
feat(ecs-service-extensions): Enable default logging to CloudWatch for extensions (under feature flag)

### PR DESCRIPTION
**Pull request by upparekh**
----
This PR adds a default `awslogs` log driver to the application container. We first check if any other observability extensions (e.g. `FirelensExtension`) have already been added to the service. It will add the log driver only if no such loggers have been enabled.

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/container.ts [hunk 1]
```diff
@@ -93,6 +108,30 @@ export class Container extends ServiceExtension {
...
+      // Create a log group for the service, if one is not provided by the user
+      if (!this.logGroup) {
+        this.logGroup = new awslogs.LogGroup(this.scope, `${this.parentService.id}-logs`, {
+          logGroupName: `${this.parentService.id}-logs`,
+          removalPolicy: cdk.RemovalPolicy.DESTROY,
+          retention: awslogs.RetentionDays.ONE_MONTH,
+        });
+      }
+
+      containerProps = {
```
Comment (iamhopaul123) [inline packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/container.ts:123]: Is this equal to `containerProps.logging = ...`?

File: packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/container.ts [hunk 2]
```diff
@@ -93,6 +108,30 @@ export class Container extends ServiceExtension {
...
+        ...containerProps,
+        logging: new ecs.AwsLogDriver({
+          streamPrefix: this.parentService.id,
+          logGroup: this.logGroup,
+        }),
+      };
+    } else {
+      if (this.logGroup) {
+        throw Error(`A log configuration has already been specified for the service '${this.parentService.id}'. The default log group provided to the service container will not be used.`);
```
Comment (iamhopaul123) [inline packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/container.ts:132]: The error message looks more like a warning to me. Maybe be more clear why it is an error like you cannot specify ... with ...

File: packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/container.ts (hunk 1 above)
Comment (upparekh) [reply packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/container.ts:123]: Yes it is! But as `logging` is a `read-only` property, we cannot update it’s value this way. So we need to create a new object with the updated values.

File: packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/container.ts (hunk 2 above)
Comment (upparekh) [reply packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/container.ts:132]: Updated!

File: packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/container.ts [hunk 3]
```diff
@@ -51,6 +60,11 @@ export class Container extends ServiceExtension {
    */
   public readonly trafficPort: number;
 
+  /**
+   * The default log group into which application container logs should be routed.
+   */
+  private _logGroup?: awslogs.ILogGroup;
```
Comment (paragbhingre) [inline packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/container.ts]: Is there any reason why we add underscore here? Or it is just to keep 2 names little different than each other?

File: packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/container.ts [hunk 4]
```diff
@@ -93,6 +108,30 @@ export class Container extends ServiceExtension {
       containerProps = hookProvider.mutateContainerDefinition(containerProps);
     });
 
+    // If no observability extensions have been added to the service description, we can configure
+    // default awslogs log driver
+    if (!containerProps.logging) {
+      // Create a log group for the service, if one is not provided by the user
+      if (!this.logGroup) {
+        this.logGroup = new awslogs.LogGroup(this.scope, `${this.parentService.id}-logs`, {
```
Comment (paragbhingre) [inline packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/container.ts:116]: Is this is a default policy configurations or best practices mentioned somewhere?

**Review by iamhopaul123 — COMMENTED** [review]
Looks good to me!

File: packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/container.ts (hunk 4 above)
Comment (upparekh) [reply packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/container.ts:116]: Actually the CDK default log group configuration would retain the logs for 2 years, it would be wasteful and would incur unnecessary costs while the most common use case for these logs would be to help in troubleshooting. I couldn't find a best practices doc but chose these values looking at precursors (like Copilot). Also, the user can provide a log group if they want to customize these values!

File: packages/@aws-cdk-containers/ecs-service-extensions/test/container.test.ts [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+          'CustomTaskRole3C6B13FD',
+          'Arn',
+        ],
+      },
+    });
+
+
+  });
+
+  test('should error when log group is provided but feature flag is not set', () => {
```
Comment (madeline-k) [inline packages/@aws-cdk-containers/ecs-service-extensions/test/container.test.ts:290]: Please add unit tests for the other error scenario as well - a logGroup is provided on the Container extension, but also a logging extension is added to the service description.

**Review by madeline-k — CHANGES_REQUESTED** [review]
The user should be able to use this feature explicitly without adding the feature flag. The feature flag should just be there to prevent breaking changes when a user upgrades, and makes no code changes. The feature flag should not block the logGroup parameter entirely from being used. Hopefully that makes sense. 

Said another way: Customers must take some action to opt-in to this behavior, it should not happen spontaneously when they upgrade. So, they can either turn on the feature flag, or set the `logGroup` property.

File: packages/@aws-cdk-containers/ecs-service-extensions/test/container.test.ts (hunk 5 above)
Comment (madeline-k) [reply packages/@aws-cdk-containers/ecs-service-extensions/test/container.test.ts:290]: oh, nevermind. I see that was in another unit test file.

File: packages/@aws-cdk-containers/ecs-service-extensions/README.md [hunk 6]
```diff
@@ -109,7 +109,39 @@ nameDescription.add(new Container({
...
+nameDescription.add(new Container({
+  cpu: 1024,
+  memoryMiB: 2048,
+  trafficPort: 80,
+  image: ContainerImage.fromRegistry('nathanpeck/name'),
+  environment: {
+    PORT: '80',
+  },
+  // Optionally provide a log group to send the application container logs to (If not provided, a log group will be created on your behalf)
+  logging: new awslogs.LogGroup(stack, 'MyLogGroup'),
```
[the remark this answers is not in the document]
Comment (upparekh) [reply packages/@aws-cdk-containers/ecs-service-extensions/README.md]: Good catch! It was actually supposed to be `logGroup`.

File: packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/container.ts [hunk 7]
```diff
@@ -51,6 +61,11 @@ export class Container extends ServiceExtension {
    */
   public readonly trafficPort: number;
 
+  /**
+   * The default log group into which application container logs should be routed.
```
[the remark this answers is not in the document]
Comment (upparekh) [reply packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/container.ts:65]: By default it means that if the user doesn't provide a log group, logging will still be enabled as default behavior for the service. But there won't be multiple log groups for an application container so I can see where it might get confusing.

File: packages/@aws-cdk-containers/ecs-service-extensions/README.md [hunk 8]
```diff
@@ -109,7 +109,38 @@ nameDescription.add(new Container({
...
+  image: ContainerImage.fromRegistry('nathanpeck/name'),
+  environment: {
+    PORT: '80',
+  },
+  // Optionally provide a log group to send the application container logs to (If not provided, a log group will be created on your behalf)
+  logGroup: new awslogs.LogGroup(stack, 'MyLogGroup'),
+}));
+```
+
+This logging behavior is supported by default if using CDK v1.140.0 or v2.6.0 and later. For enabling this behavior for previous versions, ensure that the `ECS_SERVICE_EXTENSIONS_ENABLE_DEFAULT_LOG_DRIVER` flag within the application stack context is set to true, like so:
```
Comment (madeline-k) [inline packages/@aws-cdk-containers/ecs-service-extensions/README.md:135]: This is a bit hard to describe in english rather than code 😂 . Here is my suggestion:

```suggestion
If a log group is not provided, no observability extensions have been created, and the `ECS_SERVICE_EXTENSIONS_ENABLE_DEFAULT_LOG_DRIVER` feature flag is enabled, then logging will be configured by default and a log group will be created for you. 

The `ECS_SERVICE_EXTENSIONS_ENABLE_DEFAULT_LOG_DRIVER` feature flag is enabled by default in any CDK apps that are created with CDK v1.140.0 or v2.8.0 and later.
To enable default logging for previous versions, ensure that the `ECS_SERVICE_EXTENSIONS_ENABLE_DEFAULT_LOG_DRIVER` flag within the application stack context is set to true, like so:
```

Also updated the version number - hopefully these will still be the release numbers for next week.

File: packages/@aws-cdk-containers/ecs-service-extensions/test/service.test.ts [hunk 9]
```diff
@@ -25,87 +22,6 @@ describe('service', () => {
     }).toThrow(/Service 'my-service' must have a Container extension/);
 
 
-  });
-
-  test('should be able to add a container to the service', () => {
```
Comment (madeline-k) [inline packages/@aws-cdk-containers/ecs-service-extensions/test/service.test.ts:30]: Why is this test being removed? It seems unrelated to this change.

**Review by madeline-k — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk-containers/ecs-service-extensions/test/service.test.ts (hunk 9 above)
Comment (upparekh) [reply packages/@aws-cdk-containers/ecs-service-extensions/test/service.test.ts:30]: I moved all `Container` extension related tests to a new test file `container.test.ts` (you can find this test there). I felt that as we add more properties to this extension, it would be better to have a separate test for it than to append it to a test file for another Construct.
Would you suggest I make it a new PR?

**Review by madeline-k — APPROVED** [review]
Looks great! 😄
